PagedCollection updates

tjprescott opened the pull request:
This PR:

- Fixes #303 by adding an `isExhausted` property the app developer can check.
- Wraps completion handler calls in `nextPage` and `nextItem` in appropriate calls to dispatch onto the main queue to eliminate the need for the developer to do this.
- The default behavior for "no data received" is to simply return rather than call the completion handler with `nil` data. This allows the user's app to do the right thing when the collection is exhausted without them having to manually check `isExhausted`. The property is still supplied as the app developer may want to affirmatively check this.

tjprescott commented:
@bsiegel ping :)

bsiegel commented on examples/AzureSDKDemoSwift/Source/BlobDownloadViewController.swift:
```diff
@@ -79,25 +78,20 @@ class BlobDownloadViewController: UIViewController, MSALInteractiveDelegate {
         dataSource?.nextPage { result in
             switch result {
             case let .success(data):
                 if data != nil {
```
If I read the commentary right, this condition should now always be true right?

bsiegel replied:
If that's accurate, should we change the Result type for nextPage / nextItem to have a non-nullable element in the success case?

tjprescott replied:
Good point. I've updated those return types and adjusted accordingly. Now the client code doesn't need to worry about the "got nothing" scenario at all!

bsiegel replied:
Awesome, yeah that definitely simplifies things. Probably just need to update the demo app here to match that and then this is 👍 to merge.

tjprescott commented:
@bsiegel I updated the demo app accordingly, and added a second commit that fixes a small bug in the filtering methods. Please make sure you are okay with the fix.

bsiegel commented on sdk/storage/AzureStorageBlob/Source/DataStructures/TransferCollection.swift:
```diff
@@ -96,12 +96,14 @@ public struct TransferCollection {
+        state: TransferState? = nil
     ) -> BlobTransfer? {
-        return items.first { match($0) }
+        return items.first { transfer in
```
Good catch! That's sort of the one thing I didn't like about making all of the matching arguments optional, is that the autocomplete suggests `transfers.firstWith()` as an option which obviously makes no sense, but there's no way to encode "all of these arguments are optional but at least one must be specified".

That said, I made the same mistake above in `filterWhere` - mind fixing it up there as well? Thanks <3

tjprescott replied:
Done 👍
